patches_for_WebRTC_org/m80/0007.2-Properly-handling-async-model-for-initializing-Media.patch

```diff
@@ -0,0 +1,695 @@
+-  THR(ActivateInstance(
+-      HStringReference(
+-          RuntimeClass_Windows_Media_Capture_MediaCaptureInitializationSettings)
+-          .Get(),
+-      &settings));
+-  THR(settings.As(&settings5));
+-  THR(settings5->put_MemoryPreference(
+-      MediaCaptureMemoryPreference::MediaCaptureMemoryPreference_Cpu));
+-  THR(settings->put_VideoDeviceId(device_id.Get()));
+ 
+-  THR(media_capture_->InitializeWithSettingsAsync(settings.Get(),
+-                                                  &async_action));
+-  THR(WaitForAsyncAction(async_action));
++  THR(GetMediaCaptureForDevice(device_id.Get(),
++                                media_capture_.ReleaseAndGetAddressOf()));
+ 
+ Cleanup:
+   return hr;
+@@ -183,7 +166,8 @@ HRESULT VideoCaptureWinRTInternal::StartCapture(
+   ComPtr<IMediaCapture5> media_capture5;
+   ComPtr<IMapView<HSTRING, MediaFrameSource*>> frame_sources;
+   ComPtr<IIterable<IKeyValuePair<HSTRING, MediaFrameSource*>*>> iterable;
+-  ComPtr<IIterator<IKeyValuePair<HSTRING, MediaFrameSource*>*>> frame_source_iterator;
++  ComPtr<IIterator<IKeyValuePair<HSTRING, MediaFrameSource*>*>>
++      frame_source_iterator;
+   ComPtr<IAsyncOperation<MediaFrameReader*>> async_operation;
+   ComPtr<IAsyncOperation<MediaFrameReaderStartStatus>>
+       async_media_frame_reader_start_status;
+@@ -360,9 +344,12 @@ HRESULT VideoCaptureWinRTInternal::FrameArrived(
+   ComPtr<IMemoryBuffer> memory_buffer;
+   ComPtr<IMemoryBufferReference> memory_buffer_reference;
+   ComPtr<IMemoryBufferByteAccess> memory_buffer_byte_access;
++  BitmapPlaneDescription bitmap_plane_description_y;
++  BitmapPlaneDescription bitmap_plane_description_uv;
+   HString video_subtype;
+   uint8_t* bitmap_content;
+   uint32_t bitmap_capacity;
++  int32_t plane_count;
+ 
+   THR(media_frame_reader->TryAcquireLatestFrame(&media_frame_reference));
+ 
+@@ -384,15 +371,45 @@ HRESULT VideoCaptureWinRTInternal::FrameArrived(
+     frameInfo.interlaced = false;
+ 
+     THR(video_media_frame->get_SoftwareBitmap(&software_bitmap));
++
+     THR(software_bitmap->LockBuffer(
+         BitmapBufferAccessMode::BitmapBufferAccessMode_Read, &bitmap_buffer));
++
++    THR(bitmap_buffer->GetPlaneCount(&plane_count));
++    THR(plane_count == 2 ? S_OK : E_FAIL);
++    THR(bitmap_buffer->GetPlaneDescription(0, &bitmap_plane_description_y));
++    THR(bitmap_buffer->GetPlaneDescription(1, &bitmap_plane_description_uv));
++
+     THR(bitmap_buffer.As(&memory_buffer));
+     THR(memory_buffer->CreateReference(&memory_buffer_reference));
+     THR(memory_buffer_reference.As(&memory_buffer_byte_access));
+     THR(memory_buffer_byte_access->GetBuffer(&bitmap_content,
+                                              &bitmap_capacity));
+ 
+-    pfn_incoming_frame_(bitmap_content, bitmap_capacity, frameInfo, 0);
++    // hack(auright): The following hack works around the limitations of 
++    // libWebRTC memory representations for YUV planar images. Windows::Media::Capture allows
```

Comment:
As per your second comment, a more thoughtful read of the code would lead you to ``VideoCaptureImpl::IncomingFrame``.

An overload would be another solution, but we come down to the harsh reality of project schedules and what is good enough for now. The comment in the code makes clear that this inefficiency is purposely by design and temporarily.
Would you prefer to fix this code or to invest in porting Anaheim's code over to WinRTC?

BTW, WebRTC has other representation for video frames. Check api\video\video_frame_buffer.h.

Reply:
> An overload would be another solution, but we come down to the harsh reality of project schedules and what is good enough for now.

The amount of code to convert here is roughly equivalent to the amount of code that would have been necessary to write an overload in the first place.

> BTW, WebRTC has other representation for video frames. Check api\video\video_frame_buffer.h.

The base class `VideoFrameBuffer` has a mandatory `ToI420()` which returns a `I420BufferInterface` which supports separate Y, U, and V planes with arbitrary strides. This is the base interop format for all modules to communicate with each other; every format derived needs to support that.

Reply:
@LoadLibrary 
> Would you prefer to fix this code or to invest in porting Anaheim's code over to WinRTC?

We'd prefer feature parity with WebRTC-UWP in the short term. If this code is introducing an extra copy compared to WebRTC-UWP, then it should be fixed.

Reply:
The problem wasn't related to `webrtc::I420Buffer`. I had to fix a bug in libyuv that wasn't taking YUV width as stride. Additionally, libyuv didn't have support for a gap between Y and UV planes (also fixed).